mongo/integration/unified/logger_verification.go

```diff
@@ -0,0 +1,283 @@
+	}
+
+	if len(log.LogMessages) == 0 {
+		return fmt.Errorf("log messages are required")
+	}
+
+	for _, message := range log.LogMessages {
+		if err := validateLogMessage(message); err != nil {
+			return fmt.Errorf("message is invalid: %v", err)
+		}
+	}
+
+	return nil
+}
+
+// validateExpectLogMessages will validate a slice of "clientLogMessages"
+// objects and return the first error encountered.
+func validateExpectLogMessages(logs []*clientLogMessages) error {
+	seenClientNames := make(map[string]struct{}) // Check for client duplication
+
+	for _, log := range logs {
+		if err := validateClientLogMessages(log); err != nil {
+			return fmt.Errorf("client is invalid: %v", err)
+		}
+
+		if _, ok := seenClientNames[log.Client]; ok {
+			return fmt.Errorf("duplicate client: %v", log.Client)
+		}
+
+		seenClientNames[log.Client] = struct{}{}
+	}
+
+	return nil
+}
+
+// logMessageValidator defines the expectation for log messages across all
+// clients.
+type logMessageValidator struct {
+	testCase *TestCase
+	err      chan error
+}
+
+// newLogMessageValidator will create a new "logMessageValidator" from a test
+// case.
+func newLogMessageValidator(testCase *TestCase) (*logMessageValidator, error) {
+	if testCase == nil {
+		return nil, fmt.Errorf("test case is required")
+	}
+
+	if testCase.entities == nil {
+		return nil, fmt.Errorf("entities are required")
+	}
+
+	validator := &logMessageValidator{
+		testCase: testCase,
+		err:      make(chan error, len(testCase.entities.clients())),
+	}
+
+	return validator, nil
+}
+
+type actualLogQueues map[string]chan orderedLogMessage
+
+func (validator *logMessageValidator) expected(ctx context.Context) ([]*clientLogMessages, actualLogQueues) {
+	clients := entities(ctx).clients()
+
+	expected := make([]*clientLogMessages, 0, len(validator.testCase.ExpectLogMessages))
+	actual := make(actualLogQueues, len(clients))
+
+	for _, clientLogMessages := range validator.testCase.ExpectLogMessages {
+		clientName := clientLogMessages.Client
+
+		clientEntity, ok := clients[clientName]
+		if !ok {
+			continue // If there is no entity for the client, skip it.
+		}
+
+		expected = append(expected, clientLogMessages)
+		actual[clientName] = clientEntity.logQueue
+	}
+
+	return expected, actual
+}
+
+// stopLogMessageVerificationWorkers will gracefully validate all log messages
+// received by all clients and return the first error encountered.
+func stopLogMessageVerificationWorkers(ctx context.Context, validator *logMessageValidator) error {
+	// Count the number of LogMessage objects on each ExpectedLogMessages.
+	// This will give us the number of "actual" log messages we expect to
+	// receive from each client. That is we want Σ (1 + len(messages)) for
+	// over all clients.
+	messageCard := 0
+	for _, clientLogMessages := range validator.testCase.ExpectLogMessages {
+		messageCard += len(clientLogMessages.LogMessages)
+	}
+
+	for i := 0; i < messageCard; i++ {
+		select {
+		case err := <-validator.err:
```

Comment:
We should be able to collect all messages sent to this channel by closing the `Logger.logQueue` channel, then closing all intermediating channels (i.e. all of the channels used by the log validator workers).

E.g. closing a channel pipeline: https://go.dev/play/p/nCxI5_AKTtx

Reply:
@matthewdale It turns out we need o pass the number of expected log messages to the unified test Logger for GODRIVER-2586. I am proposing that we use the context to do this. I am working on that ticket currently, would it be okay to defer this update until then?